pkg/cloudprovider/openstack.go

```diff
@@ -554,7 +617,38 @@ func (o *OpenStack) reserveNeutronIPAddress(s neutronsubnets.Subnet, ip net.IP,
 		Name:        fmt.Sprintf("egressip-%s", ip.String()),
 	}
 	p, err := neutronports.Create(o.neutronClient, opts).Extract()
+
 	if err != nil {
+		// Let's check if error suggests that port with that IP already exists in that subnet.
+		if _, ok := err.(*gophercloud.ErrDefault409); ok {
+			// If so, let's get the port and check if it's our reservation port. It's possible we've created it earlier.
+			// It's also possible user configured a taken range, we should erorr out in such case.
+			page, nestedErr := neutronports.List(o.neutronClient, neutronports.ListOpts{
+				NetworkID:   s.NetworkID,
+				FixedIPs:    []neutronports.FixedIPOpts{{SubnetID: s.ID, IPAddress: ip.String()}},
+				DeviceOwner: egressIPTag,
+				DeviceID:    generateDeviceID(serverID),
+			}).AllPages()
+			if nestedErr != nil {
+				return nil, nestedErr
+			}
+			ports, nestedErr := neutronports.ExtractPorts(page)
+			if nestedErr != nil {
+				return nil, nestedErr
+			}
+			if len(ports) > 1 {
+				// This is unexpected, Neutron should not allow multiple ports with the same IP
+				// in one subnet
+				return p, fmt.Errorf(
+					"bogus result from Neutron, multiple ports with IP %s in subnet %s", ip.String(), s.ID)
+			} else if len(ports) == 0 {
+				// Apparently it's not our port, return the original error.
+				return nil, err
+			}
+			// Otherwise we've got our port, let's return it.
+			return &ports[0], nil
+		}
+		// Regular error, just return it.
```

Comment:
This is part I'll need to test - not sure about the error gophercloud returns in such case.